include/crm/common/xml_internal.h

```diff
@@ -0,0 +1,118 @@
+                buf = (char *) "LOG CORRUPTION HAZARD";  /* we don't modify */\
+            } else if (len > 0 && (dechunk)) {                                \
+                buf[len - 1] = '\0';                                          \
+            }                                                                 \
+            if (buffer) {                                                     \
+                qb_log_from_external_source(__FUNCTION__, __FILE__, "%s%s%s", \
+                                            (priority), __LINE__, 0,          \
+                                            (prefix) != NULL ? (prefix) : "", \
+                                            buffer, buf);                     \
+                free(buffer);                                                 \
+            } else {                                                          \
+                qb_log_from_external_source(__FUNCTION__, __FILE__, "%s%s",   \
+                                            (priority), __LINE__, 0,          \
+                                            (prefix) != NULL ? (prefix) : "", \
+                                            buf);                             \
+            }                                                                 \
+            if (len < 0) {                                                    \
+                buf = NULL;  /* restore temporary override */                 \
+            }                                                                 \
+            buffer = NULL;                                                    \
+            buffer_len = 0;                                                   \
+            (void) (postemit);                                                \
+                                                                              \
+        } else if (buffer == NULL) {                                          \
+            buffer_len = len;                                                 \
+            buffer = buf;                                                     \
+            buf = NULL;                                                       \
+                                                                              \
+        } else {                                                              \
+            buffer = realloc(buffer, 1 + buffer_len + len);                   \
+            memcpy(buffer + buffer_len, buf, len);                            \
+            buffer_len += len;                                                \
+            buffer[buffer_len] = '\0';                                        \
+        }                                                                     \
+        free(buf);                                                            \
+    }                                                                         \
+} while (0);
```

Comment:
omit the final semicolon in the do...while(0) construct ... the idea is that it should work wherever a single statement would, including:

  if (A)
     CALL_MACRO();
  else
     something_else();

even though our convention is to always use braces ...